packages/pg-cursor/index.js

```diff
@@ -202,6 +206,49 @@ class Cursor extends EventEmitter {
     this.connection.end()
   }
 
+  _promisifiedQuery(text) {
+    return new Promise((resolve, reject) => {
+      try {
+        var res = this.connection.query(text)
+        resolve(res)
+      } catch (e) {
+        reject(e)
+      }
+    })
+  }
+
+  moveAllAndClose(cb) {
+    this.moveAllAndClosing = true
+    if (this.state === 'done' || this.state == 'error') {
+      return this.close(cb)
+    }
+    return this._promisifiedQuery("MOVE ALL FROM " + this._portal).then(() => {
+      this._close_checkFinished(cb)
+    })
+  }
+
+  _close_checkFinished(cb) {
+    this._closePortal()
+    // try to delay the callback until portal is finished closing
+    // (this means a closeComplete followed by a readyForQuery)
+    // (it's possible to have a readyForQuery without a closeComplete if called from moveAllAndClose)
```

Comment:
Is the order of messages as follows:
```
Parse query as foo->
<-ParseComplete
Bind foo to bar->
<-BindComplete
Execute bar->
<-PortalSuspended|ErrorResponse|EmptyQueryResponse|CommandComplete
...[repeat as desired]...
Execute bar->
<-PortalSuspended|ErrorResponse|EmptyQueryResponse|CommandComplete
Query "MOVE ALL bar"->
<-CommandComplete|ErrorResponse
<-ReadyForQuery
Close bar->
<-CloseComplete|ErrorResponse
Sync->
<-ReadyForQuery
```
?

If so, it what case should we not expect a CloseComplete after calling this._closePortal? An ErrorResponse from the MOVE ALL? And if it's the case we won't get a CloseComplete, then is it possible we'll never get to line 241?

Reply:
I think you always get a CloseComplete, the issue has to do with the ReadyForQuery responses. 

The issue I was trying to solve here is that the MOVE ALL query gives an extra ReadyForQuery, so the callback cb in close(cb) is executed before the close() itself gives a ReadyForQuery. That is, the close logic doesn't care about where the ReadyForQuery is coming from, so it'll execute prematurely on the move-all-readyforquery instead of the close-portal-readyforquery. This means that sometimes the portal-closing ReadyForQuery will get sent as the next cursor is being made, and that extra ReadyForQuery causes an error in the new cursor. 

So the close complete will always be sent, this is just my solution for ensuring that the callback is delayed long enough for everything to execute properly.